File: src/lib/__tests__/resolve.test.ts
import { describe, it, expect } from "vitest";
import { CircleHelp, Lock } from "lucide-react";
import {
  resolveName,
  resolveIcon,
  prettifyEntityId,
} from "../resolve";
import { overrides } from "../../config/overrides";
import type { HassEntity } from "../ha";

const make = (
  entity_id: string,
  friendly_name?: string,
  extra: Record<string, unknown> = {},
): HassEntity => ({
  entity_id,
  state: "on",
  attributes: { ...(friendly_name ? { friendly_name } : {}), ...extra },
});

describe("resolveName", () => {
  it("override wins over friendly_name and id (kills 'Lock Current status …')", () => {
    const contact = make(
      "binary_sensor.front_door_current_status",
      "Lock Current status",
    );
    expect(resolveName(contact, overrides)).toBe("Front Door");
  });

  it("falls back to friendly_name when no override", () => {
    const e = make("sensor.kitchen_temp", "Kitchen Temperature");
    expect(resolveName(e, overrides)).toBe("Kitchen Temperature");
  });

  it("prettifies the entity_id as last resort", () => {
    const e = make("switch.garage_work_light");
    expect(resolveName(e, {})).toBe("Garage Work Light");
  });

  it("ignores blank friendly_name", () => {
    const e = make("light.hall", "   ");
    expect(resolveName(e, {})).toBe("Hall");
  });
});

describe("prettifyEntityId", () => {
  it("strips the domain and title-cases", () => {
    expect(prettifyEntityId("binary_sensor.front_door_current_status")).toBe(
      "Front Door Current Status",
    );
  });
});

describe("resolveIcon", () => {
  it("uses the override icon when present", () => {
    const e = make("lock.front_door_lock", "Lock");
    expect(resolveIcon(e, overrides)).toBe(Lock);
  });

  it("falls back to a neutral icon for an unknown domain (never throws)", () => {
    const e = make("totally_made_up.thing");
    expect(() => resolveIcon(e, {})).not.toThrow();
    expect(resolveIcon(e, {})).toBe(CircleHelp);
  });
});

File: android/app/src/test/java/com/hawksnest/core/logic/ResolveTest.kt
package com.hawksnest.core.logic

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.QuestionMark
import com.hawksnest.config.overrides
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertSame

/** Ported from `src/lib/__tests__/resolve.test.ts`. */
class ResolveTest {

    @Test
    fun `override wins over friendly_name and id`() {
        val contact = entity("binary_sensor.front_door_current_status", "Lock Current status")
        assertEquals("Front Door", resolveName(contact, overrides))
    }

    @Test
    fun `falls back to friendly_name when no override`() {
        val e = entity("sensor.kitchen_temp", "Kitchen Temperature")
        assertEquals("Kitchen Temperature", resolveName(e, overrides))
    }

    @Test
    fun `prettifies the entity_id as last resort`() {
        assertEquals("Garage Work Light", resolveName(entity("switch.garage_work_light")))
    }

    @Test
    fun `ignores blank friendly_name`() {
        assertEquals("Hall", resolveName(entity("light.hall", "   ")))
    }

    @Test
    fun `prettifyEntityId strips the domain and title-cases`() {
        assertEquals(
            "Front Door Current Status",
            prettifyEntityId("binary_sensor.front_door_current_status"),
        )
    }

    @Test
    fun `resolveIcon uses the override icon when present`() {
        val e = entity("lock.front_door_lock", "Lock")
        assertSame(Icons.Filled.Lock, resolveIcon(e, overrides))
    }

    @Test
    fun `resolveIcon falls back to a neutral icon for an unknown domain`() {
        val e = entity("totally_made_up.thing")
        assertSame(Icons.Filled.QuestionMark, resolveIcon(e))
    }
}
